### ISSUE TITLE
clang and clang-tidy doesn't understand gcc's -fno-canonical-system-headers

### ISSUE THREAD

**Issue by matta** [clangd/clangd#1004]
Please describe the problem.

Google's Bazel uses gcc as the system compiler.  I've been using https://github.com/hedronvision/bazel-compile-commands-extractor to get a `compile_commands.json`.

Clang tooling recognizes almost all aspects of these command lines, except that bazel passes `-fno-canonical-system-headers` to gcc, which clang tooling doesn't recognize.

I can work around this in `.clangd` with `CompileFlags/Remove`, but this doesn't translate to other clang tools run in the same tree.

``` sh
$ clang-tidy testcode.cxx
52802 warnings and 1 error generated.
Error while processing /home/matt/git/binary-search-trees/base/testcode.cxx.
warning: redefining builtin macro [clang-diagnostic-builtin-macro-redefined]
error: unknown argument: '-fno-canonical-system-headers' [clang-diagnostic-error]
Suppressed 52799 warnings (52799 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler error(s).
```
So, it'd be nice if all clang tooling ignored `-fno-canonical-system-headers` across the board.  If this isn't the right place to file such an issue, let me know where I can direct this.

**Logs**

Please attach the clangd stderr log if you can. (Usually available from the editor)
If possible, run with `--log=verbose` - note that the logs will include the contents of open files!
If this is a crash, try to put `llvm-symbolizer` on your PATH per the troubleshooting instructions.

**System information**

Output of `clangd --version`:

Debian clangd version 13.0.1-+rc1-1~exp4
Features: linux
Platform: x86_64-pc-linux-gnu

Editor/LSP plugin: Emacs w/eglot

Operating system: Debian Testing

**Comment by HighCommander4** [comment]
> So, it'd be nice if all clang tooling ignored `-fno-canonical-system-headers` across the board. If this isn't the right place to file such an issue, let me know where I can direct this.

Since you're looking to get this work with all types of clang tooling, it would make sense to file an upstream issue at https://github.com/llvm/llvm-project/issues (probably with the label `clang:driver`).

**Comment by davehodg** [comment]
I just ran into this. "bazel clean --expunge" fixed it.

**Comment by cpsauer** [comment]
@matta, did you get a chance to file this upstream?

**Comment by matta** [comment]
> @matta, did you get a chance to file this upstream?

@cpsauer, nope. I'm no longer working with this stuff so I'm no longer a good person to do so.

**Comment by cpsauer** [comment]
Thanks all. Filed upstream at https://github.com/llvm/llvm-project/issues/61699

**Comment by TamaMcGlinn** [comment]
To elaborate on this

> I can work around this in .clangd with CompileFlags/Remove,

that means putting a file named .clangd in some parent directory of the working directory with this in it:

```
CompileFlags:
  Remove:
    - -fno-canonical-system-headers
```
